Augment lease revocation metrics (#11954)
hashicorp/vault: website/content/docs/internals/telemetry.mdx

| `vault.expire.job_manager.queue_length`                                                         | Total pending revocation jobs by auth method                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | leases   | sample  |

> | `vault.expire.job_manager.workers` (cluster, queue_id)                                          | Count of expiration workers per auth mount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | leases   | counter |

Author: couple things to point out for reviewers
1. Is `queue_id` the correct term to use here? ( as opposed to `mount_point`, `auth_method` or `mount_id` )
2. Is `leases` the correct unit ? Technically it should be workers, but that won't mean much externally, so not sure.

Author: removed in 1df235f

| `vault.expire.num_leases`                                                                       | Number of all leases which are eligible for eventual expiry                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | leases   | gauge   |

> | `vault.expire.num_irrevocable_leases`                                                           | Number of irrevocable leases which are present in the system                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | leases   | gauge   |

Reviewer: ```suggestion
| `vault.expire.num_irrevocable_leases`                                                           | Number of irrevocable leases that cannot be revoked automatically                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | leases   | gauge   |
```

this phrasing might be more clear? if you're ever in doubt about phrasing, Ivana is a great resource!

Author: ah nice! yeah I'm gonna use this wording, thanks!

Reviewer: irrevocable is a bit redundant in this case, but probably nice for a user CTL/CMD-F-ing through the documentation

Author: Ah yeah I was thinking of getting rid of `irrevocable` here, making it `Number of leases that cannot be revoked automatically`, and since that is in the metric name itself, users should still be able to find it on searching i think.

Reviewer: yeah true, sounds good!

Author: 2558e12

| `vault.expire.lease_expiration.error`                                                           | Count of lease expiration errors                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | errors   | counter |

> | `vault.expire.queue_workers` (cluster, queue_id)                                                | Count of expiration workers per auth mount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | leases   | counter |

Reviewer: if you update the naming be sure to update here. the form for the suggestion I gave would be `vault.expire.job_manager.workers`. I believe what you currently have would lead to `vault.expire.expire.queue_workers` (I think `expire` is prefixed to job manager metrics that are called by the expiration manager

Reviewer: now that I'm looking at it, I'm not sure that the metric sink is prefixed with `expire` since we're passing in the one from core. so perhaps `vault.expire.job_manager.queue_length` would have actually been `vault.job_manager.queue_length`. Let's look into this.

Author: ah so I think the metricSink struct is not prefixed with `expire` itself, but when emitting the metrics this is done by starting the metric name with `j.name` https://github.com/hashicorp/vault/blob/main/helper/fairshare/jobmanager.go#L120 which is `expire` in the case of the jobmanager used by expiration.go.
So once I update this new metric to be on par with those, it should then come out to `vault.expire.job_manager.workers` ( also changing the `queue_workers` back to `workers`, cause externally nobody would understand what this internal queue is for anyways so guess it'd be confusing )

Author: ah looks like I can't make a different comment on this line so just gonna reuse this. I have 2 doubts on this line
1. Should the unit of the metric be `leases` ? I'm thinking it should be `workers`, but that wouldn't mean much for external people, so not sure what a more external-friendly unit should be here.
2. In the labels for this metric, I've used `queue_id`. This could just as easily be `auth_method` or `mount_point` or `mount_id`, based on other words we've used in the telemetry docs, i believe. So I'm not sure if `queue_id` is the right choice here.
Note: The metric `vault.expire.job_manager.queue_length` on L157 seems to be missing a note on the label, as it also uses a `queue_id` label in the code.

Reviewer 2: I worry about the cardinality of metrics that have a label based on mounts.  There are customers out there with 1000s of mounts, and most metrics systems don't cope well with cardinality over ~100.  We do have some per-mount metrics, but those I believe are constrained by settings like https://www.vaultproject.io/docs/configuration/telemetry#usage_gauge_period and https://www.vaultproject.io/docs/configuration/telemetry#maximum_gauge_cardinality.

Author: that concern makes a lotta sense! I did think briefly about emitting this metric as a gauge (off of the workerCount map) to reduce the volume of metrics emission, but that ended up missing workers that would do their job and quit in between metric emissions, plus it wouldn't address the cardinality issue. 
So doing this as a gauge would probably limit the effectiveness of this metric. In that case, would it just make more sense to leave this out for now, and only add it in a later release if there is customer demand for it, so the situation is more `can we please have this` versus `why is this blowing up my grafana budget` ?

Author: Fixed up the metric name here 2558e12, although the question around whether we should have this metric, and the couple doubts on language are still open.

Author: Aha, did not notice that, good spot!
Roger that, will go ahead with this for now! Btw, I'm gonna be working on https://github.com/hashicorp/vault/issues/3792 soon as my current plate empties a little, think this might be helpful for customers to just turn off metrics if they aren't gaining value from them.
